src/file_pipe_log.rs

```diff
@@ -0,0 +1,959 @@
+        let path = dir.path().to_str().unwrap();
+
+        let rotate_size = 1024;
+        let bytes_per_sync = 32 * 1024;
+        let (pipe_log, _) = new_test_pipe_log(path, bytes_per_sync, rotate_size);
+        assert_eq!(pipe_log.first_file_id(queue), INIT_FILE_NUM.into());
+        assert_eq!(pipe_log.active_file_id(queue), INIT_FILE_NUM.into());
+
+        let header_size = (FILE_MAGIC_HEADER.len() + VERSION.len()) as u64;
+
+        // generate file 1, 2, 3
+        let content: Vec<u8> = vec![b'a'; 1024];
+        let (file_num, offset, _) = pipe_log.append_bytes(queue, &content, &mut false).unwrap();
+        assert_eq!(file_num, 1.into());
+        assert_eq!(offset, header_size);
+        assert_eq!(pipe_log.active_file_id(queue), 1.into());
+
+        let (file_num, offset, _) = pipe_log.append_bytes(queue, &content, &mut false).unwrap();
+        assert_eq!(file_num, 2.into());
+        assert_eq!(offset, header_size);
+        assert_eq!(pipe_log.active_file_id(queue), 2.into());
+
+        // purge file 1
+        assert_eq!(pipe_log.purge_to(queue, 2.into()).unwrap(), 1);
+        assert_eq!(pipe_log.first_file_id(queue), 2.into());
+
+        // cannot purge active file
+        assert!(pipe_log.purge_to(queue, 3.into()).is_err());
+
+        // append position
+        let s_content = b"short content".to_vec();
+        let (file_num, offset, _) = pipe_log
+            .append_bytes(queue, &s_content, &mut false)
+            .unwrap();
+        assert_eq!(file_num, 3.into());
+        assert_eq!(offset, header_size);
+
+        let (file_num, offset, _) = pipe_log
+            .append_bytes(queue, &s_content, &mut false)
+            .unwrap();
+        assert_eq!(file_num, 3.into());
+        assert_eq!(offset, header_size + s_content.len() as u64);
+
+        assert_eq!(
+            pipe_log.active_log_size(queue),
+            FILE_MAGIC_HEADER.len() + VERSION.len() + 2 * s_content.len()
+        );
+
+        let content_readed = pipe_log
+            .read_bytes(queue, 3.into(), header_size, s_content.len() as u64)
+            .unwrap();
+        assert_eq!(content_readed, s_content);
+
+        // leave only 1 file to truncate
+        assert!(pipe_log.purge_to(queue, 3.into()).is_ok());
+        assert_eq!(pipe_log.first_file_id(queue), 3.into());
+        assert_eq!(pipe_log.active_file_id(queue), 3.into());
+
+        // truncate file
+        pipe_log
+            .truncate_active_log(queue, Some(FILE_MAGIC_HEADER.len() + VERSION.len()))
+            .unwrap();
+        assert_eq!(
+            pipe_log.active_log_size(queue,),
+            FILE_MAGIC_HEADER.len() + VERSION.len()
+        );
+        let trunc_big_offset = pipe_log.truncate_active_log(
+            queue,
+            Some(FILE_MAGIC_HEADER.len() + VERSION.len() + s_content.len()),
+        );
+        assert!(trunc_big_offset.is_err());
+
+        /**************************
+        // read next file
```

Comment:
Seems it  isn't commentted in the origin file?

Reply:
It is, see https://github.com/tikv/raft-engine/blob/master/src/pipe_log.rs#L857-L865